src/vanilla.ts

```diff
@@ -45,44 +25,50 @@ export type StoreApi<T extends State> = {
   subscribe: Subscribe<T>
   destroy: Destroy
 }
+
 export type StateCreator<
   T extends State,
-  CustomSetState = SetState<T>,
-  CustomGetState = GetState<T>,
-  CustomStoreApi extends StoreApi<T> = StoreApi<T>
-> = (set: CustomSetState, get: CustomGetState, api: CustomStoreApi) => T
-
-function createStore<
-  TState extends State,
-  CustomSetState,
-  CustomGetState,
-  CustomStoreApi extends StoreApi<TState>
->(
-  createState: StateCreator<
-    TState,
-    CustomSetState,
-    CustomGetState,
-    CustomStoreApi
-  >
-): CustomStoreApi
-
-function createStore<TState extends State>(
-  createState: StateCreator<TState, SetState<TState>, GetState<TState>, any>
-): StoreApi<TState>
-
-function createStore<
-  TState extends State,
-  CustomSetState,
-  CustomGetState,
-  CustomStoreApi extends StoreApi<TState>
+  Mis extends [StoreMutatorIdentifier, unknown][],
+  Mos extends [StoreMutatorIdentifier, unknown][],
+  U = T
+> = ((
+  setState: Get<Mutate<StoreApi<T>, Mis>, 'setState', undefined>,
+  getState: Get<Mutate<StoreApi<T>, Mis>, 'getState', undefined>,
+  store: Mutate<StoreApi<T>, Mis>,
+  $$storeMutations: Mis
+) => U) & { $$storeMutators?: Mos }
+
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+export interface StoreMutators<S, A> {}
+export type StoreMutatorIdentifier = keyof StoreMutators<unknown, unknown>
+
+export type Mutate<S, Ms> = Ms extends []
+  ? S
+  : Ms extends [[infer Mi, infer Ma], ...infer Mrs]
+  ? Mutate<StoreMutators<S, Ma>[Mi & StoreMutatorIdentifier], Mrs>
+  : never
+
+type Get<T, K, F = never> = K extends keyof T ? T[K] : F
+
+type CreateStore = {
+  <T extends State, Mos extends [StoreMutatorIdentifier, unknown][] = []>(
+    initializer: StateCreator<T, [], Mos>
+  ): Mutate<StoreApi<T>, Mos>
+
+  <T extends State>(): <Mos extends [StoreMutatorIdentifier, unknown][] = []>(
+    initializer: StateCreator<T, [], Mos>
+  ) => Mutate<StoreApi<T>, Mos>
+}
+
+type CreateStoreImpl = <
+  T extends State,
+  Mos extends [StoreMutatorIdentifier, unknown][] = []
 >(
-  createState: StateCreator<
-    TState,
-    CustomSetState,
-    CustomGetState,
-    CustomStoreApi
-  >
-): CustomStoreApi {
+  initializer: StateCreator<T, [], Mos>
+) => Mutate<StoreApi<T>, Mos>
+
+const _createStore: CreateStoreImpl = (createState) => {
```

Comment:
```suggestion
const createStoreImpl: CreateStoreImpl = (createState) => {
```

I prefer this naming. (In TS, `_foo` is used for "unused", so prefer avoiding "internal" meaning.)

Reply:
Yeah I think I prefer that too because `fooImpl` is also how other middlewares are implemented